core/src/main/java/org/apache/calcite/rel/rules/ReduceExpressionsRule.java

```diff
@@ -868,8 +872,8 @@ protected static RexNode substitute(RexCall call, int ordinal, RexNode node) {
     }
 
     private RexNode visit(final RexNode call) {
-      int i = reducibleExps.indexOf(call);
-      if (i == -1) {
+      Integer i = reducibleExpsMap.get(call);
+      if (i == null) {
         return null;
       }
       RexNode replacement = reducedValues.get(i);
```

Comment:
Depending on the implementation of List, reducedValues.get(i) and addCasts.get(i) can be O(n) complexity as well. Right now we are using ArrayList, which is fine. But could be a problem potentially in the future.

Reply:
+1, better to update it too.

Reply:
`reducedValues` is not a local variable, and it would be a non-trival task to refactor this code. So I suggest we merge the PR as is.